### ISSUE TITLE
为什么当我加大batch_size后训练速度反而增大了呢

### ISSUE THREAD

**Issue by dudu8246** [THUDM/ChatGLM2-6B#385]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

如题，加大batch_size后 几张显卡都没有占满，batch_size从1增加到8 都会增加几倍的时长 ，正常不是应该会加快训练速度吗

### Steps To Reproduce

PRE_SEQ_LEN=128
LR=2e-2
NUM_GPUS=5

torchrun --standalone --nnodes=1 --nproc-per-node=$NUM_GPUS main.py \
    --do_train \
    --train_file AdvertiseGen/train.json \
    --validation_file AdvertiseGen/dev.json \
    --preprocessing_num_workers 64 \
    --prompt_column content \
    --response_column summary \
    --history_column history \
    --overwrite_cache \
    --model_name_or_path /home/chatglm2-6b-git \
    --output_dir /home/chatglm2-pt-chat-$PRE_SEQ_LEN-$LR \
    --overwrite_output_dir \
    --max_source_length 256 \
    --max_target_length 256 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 16 \
    --predict_with_generate \
    --max_steps 3000 \
    --logging_steps 10 \
    --save_steps 1000 \
    --learning_rate $LR \
    --pre_seq_len $PRE_SEQ_LEN

### Environment

```markdown
- OS:
- Python:
- Transformers:
- PyTorch:
- CUDA Support (`python -c "import torch; print(torch.cuda.is_available())"`) :
```

**Comment by XavierYu404** [comment]
--gradient_accumulation_steps 16 把这项调小就可以变快点

**Comment by dudu8246** [comment]
> 

感谢指点，不过在这个参数不变的情况下  正常适量的加大batch_size 不是应该加速训练吗  我的GPU还有很多剩余 我试了几次 基本上每次加大size都会 线性的增加训练时间。。

**Comment by stvea** [comment]
建议减小Gradient accumulation的设置，因为它是一个用速度换显存的操作。当你增大Batch Size的时候，这个操作的耗时会被放大。
具体可以参考[ Efficient Training on a Single GPU](https://huggingface.co/docs/transformers/perf_train_gpu_one#efficient-training-on-a-single-gpu)

> Gradient Accumulation
> The idea behind gradient accumulation is to instead of calculating the gradients for the whole batch at once to do it in smaller steps. The way we do that is to calculate the gradients iteratively in smaller batches by doing a forward and backward pass through the model and accumulating the gradients in the process. When enough gradients are accumulated we run the model’s optimization step. This way we can easily increase the overall batch size to numbers that would never fit into the GPU’s memory. In turn, however, the added forward and backward passes can slow down the training a bit.
>

**Comment by dudu8246** [comment]
> 建议减小Gradient accumulation的设置，因为它是一个用速度换显存的操作。当你增大Batch Size的时候，这个操作的耗时会被放大。 具体可以参考[ Efficient Training on a Single GPU](https://huggingface.co/docs/transformers/perf_train_gpu_one#efficient-training-on-a-single-gpu)
> 
> > Gradient Accumulation
> > The idea behind gradient accumulation is to instead of calculating the gradients for the whole batch at once to do it in smaller steps. The way we do that is to calculate the gradients iteratively in smaller batches by doing a forward and backward pass through the model and accumulating the gradients in the process. When enough gradients are accumulated we run the model’s optimization step. This way we can easily increase the overall batch size to numbers that would never fit into the GPU’s memory. In turn, however, the added forward and backward passes can slow down the training a bit.
我明白了  非常感谢

**Comment by dudu8246** [comment]
> > 建议减少梯度累加的设置，因为它是一个用速度换显存的操作。当你增大Batch Size的时候，这个操作的运行时间会被放大。具体可以参考Efficient Training on a Single [GPU](https://huggingface.co/docs/transformers/perf_train_gpu_one#efficient-training-on-a-single-gpu)
> > > 梯度累积 梯度累积
> > > 背后的想法是不是一次性计算整个批次的梯度，而是以更小的步骤进行。我们这样做的方法是通过向前和向后遍历模型并在此过程中累积梯度，以较小的批次迭代计算梯度。当积累了足够的梯度时，我们运行模型的优化步骤。通过这种方式，我们可以轻松地将总体批处理大小增加到 GPU 内存永远无法容纳的数字。然而，反过来，增加的向前和向后传球可能会稍微减慢训练速度。
> > > 我明白了，非常感谢

我刚刚按照这个方法做了实验，把梯度累加从16>1 batch_size 1>16 时间确实减少了一点点，
不过当我在梯度累加不变还是1的情况下  加大 batch_size加大到了32  发现估算训练时间还是增加了一倍（GPU显存还是有一些剩余）
我开始怀疑加大batch_size 是否真的能够加速训练时间了。。

**Comment by dudu8246** [comment]
我应该找到问题所在了，train.sh 中 设置了max_steps =3000  这导致了我们自己手动设置的epoch是失效的 ，也就是说epoch是由你的Num examples /(batch_size*Gradient Accumulation steps*num gpu) 决定的 那么在epoch没法设定上限的情况下  无论你加大多少batch_size  不oom的情况下  时间肯定会越来越高的，个人认为是原因。。
